### PR TITLE
fix(backend): carry marketplace name/description/image to downloaded library agents

### PR DESCRIPTION
**Pull request by Abhi1992002**
### Why / What / How

**Why:** When you add an agent from the marketplace to your library, it shows up with the creator's *original* graph title and description — not the title/description it was published under — and with no image. The intended behaviour (OPEN-2478 / #9879) is that a downloaded agent appears in your library just as it does in the marketplace.

**What:** Snapshot the marketplace listing's published name, description and image onto the `LibraryAgent` at download time, and surface them in the library.

**How:** `LibraryAgent` already had an `imageUrl` column that was never populated on marketplace downloads. This PR follows that same snapshot-at-download pattern:

- Adds nullable `name` / `description` columns to `LibraryAgent` (migration included). Null for user-created agents, which keep falling back to the graph's own values.
- `add_graph_to_library` now reads the `StoreListingVersion` and writes its `name`, `description` and first `imageUrls` entry onto the `LibraryAgent` — on both the create and the restore-existing (soft-deleted) paths, so re-downloading also refreshes the snapshot.
- `LibraryAgent.from_db` prefers the snapshotted `name`/`description`, falling back to `graph.name`/`graph.description`.

This is scoped to the "Add to Library" download path; user-created agents (via `create_library_agent`) are unaffected.

### Changes 🏗️

- `schema.prisma`: add `name` / `description` (both `String?`) to `LibraryAgent`; new migration `20260727120000_add_library_agent_marketplace_metadata`.
- `_add_to_library.py`: new `_get_marketplace_metadata()` helper; `add_graph_to_library` snapshots name/description/imageUrl on create and update.
- `library/model.py`: `from_db` uses `agent.name or graph.name` and `agent.description or graph.description`.
- Tests: `_add_to_library_test.py` (snapshot written on create + restore; helper edge cases) and `model_test.py` (override precedence + fallback).

#### For code changes:
- [x] I have clearly listed my changes in the PR description
- [x] I have made a test plan
- [x] I have tested my changes according to the test plan:
  - [x] `poetry run pytest backend/api/features/library/_add_to_library_test.py backend/api/features/library/model_test.py` — 9 passed
  - [x] Verified the `from_db` fallback path keeps user-created agents on their graph name/description
  - [ ] Publish an agent to the marketplace under a different title/description + image, add it to a fresh account's library, confirm the library card shows the published title/description/image

### CODE REVIEW

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py [hunk 1]
```diff
@@ -55,6 +55,28 @@ async def resolve_graph_for_library(
...
+    store_listing_version_id: str,
+) -> dict[str, str | None]:
+    """Snapshot the marketplace listing's title/description/image.
+
+    Returns the published ``name``, ``description`` and first image URL so a
+    downloaded agent shows up in the library exactly as it appears in the
+    marketplace. Missing values fall back to ``None``, in which case the graph's
+    own values are used at read time.
+    """
+    slv = await prisma.models.StoreListingVersion.prisma().find_unique(
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py]: Not an exposure introduced by this PR. The non-admin path resolves graph access via `get_graph(user_id=...)`, which grants a non-owner access only through a `StoreListingVersion` with `submissionStatus == APPROVED` and `isDeleted == False`, gated on the same graph version (backend/data/graph.py). A pending/rejected/deleted listing therefore raises `NotFoundError` before any snapshot happens. This PR does not change that lookup or authorization — it only additionally reads cosmetic name/description/image fields from the SLV that `resolve_graph_for_library` already fetches. Resolving as a false positive for this change.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py [hunk 2]
```diff
@@ -55,6 +55,28 @@ async def resolve_graph_for_library(
...
+        where={"id": store_listing_version_id}
+    )
+    if not slv:
+        return {"name": None, "description": None, "imageUrl": None}
+    return {
+        "name": slv.name,
+        "description": slv.description,
+        "imageUrl": slv.imageUrls[0] if slv.imageUrls else None,
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py]: `imageUrls` are creator-supplied and already stored on `StoreListingVersion` and rendered as `` throughout the marketplace UI today; snapshotting the same URL onto the library card adds no new sink. A trusted-origin/HTTPS allowlist belongs at submission time and would be a separate, broader change — out of scope for this fix.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py]: Fixed in 91a828f0f: `add_graph_to_library` now receives the `StoreListingVersion` that `resolve_graph_for_library` already fetched (and existence-validated), so there's no second `StoreListingVersion.find_unique`. This also restores `db_test.py::test_add_agent_to_library`'s single-fetch assertion.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py [hunk 3]
```diff
@@ -104,6 +131,9 @@ async def add_graph_to_library(
                 "isDeleted": False,
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py:132]: Addressed in 91a828f0f: `add_graph_to_library` now snapshots from the `StoreListingVersion` resolved (and existence-guaranteed) by `resolve_graph_for_library`, and `name`/`description` are non-nullable columns, so the restore/update path can no longer overwrite them with NULL. `imageUrl` is intentionally refreshed to reflect the current listing (null only when the listing genuinely has no image).

File: autogpt_platform/backend/backend/api/features/library/model.py [hunk 4]
```diff
@@ -349,8 +349,10 @@ def from_db(
             created_at=created_at,
             updated_at=updated_at,
             last_run_at=agent.lastRunAt,
-            name=graph.name,
-            description=graph.description,
+            # Prefer the marketplace title/description snapshotted at download
+            # time; fall back to the graph's own values for user-created agents.
+            name=agent.name or graph.name,
+            description=agent.description or graph.description,
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/model.py]: Fixed in 91a828f0f: `from_db` now uses `x if x is not None else graph.x`, so an intentionally-empty published name/description is preserved rather than replaced by the graph value. Added `test_from_db_preserves_empty_marketplace_description`.

File: autogpt_platform/backend/migrations/20260727120000_add_library_agent_marketplace_metadata/migration.sql [hunk 5]
```diff
@@ -0,0 +1,7 @@
+-- Snapshot the marketplace-published title/description onto the LibraryAgent at
+-- download time so a downloaded agent appears in the library as it does in the
+-- marketplace, instead of showing the creator's original graph name/description.
+-- Nullable: user-created agents keep NULL and fall back to the graph's values.
+-- AlterTable
+ALTER TABLE "LibraryAgent" ADD COLUMN "name" TEXT;
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/migrations/20260727120000_add_library_agent_marketplace_metadata/migration.sql:6]: Intentional: existing marketplace-downloaded rows keep NULL and fall back to the graph name/description (i.e. current behavior). `LibraryAgent` has no stored reference to the `StoreListingVersion` it was downloaded from, so a reliable backfill isn't possible; re-downloading (or the restore path) refreshes the snapshot.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py [hunk 6]
```diff
@@ -55,6 +55,28 @@ async def resolve_graph_for_library(
...
 
+async def _get_marketplace_metadata(
+    store_listing_version_id: str,
+) -> dict[str, str | None]:
+    """Snapshot the marketplace listing's title/description/image.
+
+    Returns the published ``name``, ``description`` and first image URL so a
+    downloaded agent shows up in the library exactly as it appears in the
+    marketplace. Missing values fall back to ``None``, in which case the graph's
+    own values are used at read time.
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py]: Fixed in 91a828f0f: `add_graph_to_library` now receives the `StoreListingVersion` that `resolve_graph_for_library` already fetched (and existence-validated), so there's no second `StoreListingVersion.find_unique`. This also restores `db_test.py::test_add_agent_to_library`'s single-fetch assertion.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py [hunk 7]
```diff
@@ -55,6 +55,28 @@ async def resolve_graph_for_library(
...
+        where={"id": store_listing_version_id}
+    )
+    if not slv:
+        return {"name": None, "description": None, "imageUrl": None}
+    return {
+        "name": slv.name,
+        "description": slv.description,
+        "imageUrl": slv.imageUrls[0] if slv.imageUrls else None,
+    }
+
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py:79]: Addressed in 91a828f0f: added `test_marketplace_metadata_without_images_yields_null_image` covering the empty-`imageUrls` branch.

File: autogpt_platform/backend/backend/api/features/library/model_test.py [hunk 8]
```diff
@@ -41,6 +45,29 @@ def _make_library_agent(
     )
 
 
+def test_from_db_prefers_marketplace_name_and_description():
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/model_test.py:48]: Addressed in 91a828f0f: `model_test.py` now covers name/description both-set, both-null, and the empty-string-preservation case.

File: autogpt_platform/backend/backend/api/features/library/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/model.py]: Fixed in 91a828f0f: `from_db` now uses `x if x is not None else graph.x`, so an intentionally-empty published name/description is preserved rather than replaced by the graph value. Added `test_from_db_preserves_empty_marketplace_description`.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py [hunk 9]
```diff
@@ -55,6 +55,28 @@ async def resolve_graph_for_library(
     return graph_model
 
 
+async def _get_marketplace_metadata(
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py]: Kept as a small internal helper whose three keys are consumed immediately by its single caller. A TypedDict/dataclass would be heavier than warranted at this scope.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py]: The public parameter is now `store_listing_version` (fully spelled). `slv` remains only as a short local inside `resolve_graph_for_library`, consistent with the surrounding code in that module.

File: autogpt_platform/backend/backend/api/features/library/model.py [hunk 10]
```diff
@@ -349,8 +349,10 @@ def from_db(
             created_at=created_at,
             updated_at=updated_at,
             last_run_at=agent.lastRunAt,
-            name=graph.name,
-            description=graph.description,
+            # Prefer the marketplace title/description snapshotted at download
+            # time; fall back to the graph's own values for user-created agents.
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/model.py]: Fixed in 91a828f0f: `from_db` now uses `x if x is not None else graph.x`, so an intentionally-empty published name/description is preserved rather than replaced by the graph value. Added `test_from_db_preserves_empty_marketplace_description`.

File: autogpt_platform/backend/backend/api/features/library/model.py [hunk 11]
```diff
@@ -349,8 +349,10 @@ def from_db(
             created_at=created_at,
             updated_at=updated_at,
             last_run_at=agent.lastRunAt,
-            name=graph.name,
-            description=graph.description,
+            # Prefer the marketplace title/description snapshotted at download
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/model.py:352]: Fixed in 91a828f0f: `from_db` now uses `x if x is not None else graph.x`, so an intentionally-empty published name/description is preserved rather than replaced by the graph value. Added `test_from_db_preserves_empty_marketplace_description`.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py]: Fixed in 91a828f0f: `add_graph_to_library` now receives the `StoreListingVersion` that `resolve_graph_for_library` already fetched (and existence-validated), so there's no second `StoreListingVersion.find_unique`. This also restores `db_test.py::test_add_agent_to_library`'s single-fetch assertion.

File: autogpt_platform/backend/backend/api/features/library/model.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/model.py]: Fixed in 91a828f0f: `from_db` now uses `x if x is not None else graph.x`, so an intentionally-empty published name/description is preserved rather than replaced by the graph value. Added `test_from_db_preserves_empty_marketplace_description`.

File: autogpt_platform/backend/backend/api/features/library/model.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/model.py:352]: Fixed in cd4d995e0: `list_library_agents` search now also matches the snapshotted `LibraryAgent.name`/`description`, so the displayed marketplace title is searchable. Added `test_list_library_agents_search_matches_snapshot_and_graph`.

File: autogpt_platform/backend/backend/api/features/library/model.py [hunk 12]
```diff
@@ -349,8 +349,16 @@ def from_db(
...
+            # time; fall back to the graph's own values for user-created agents
+            # (which have no snapshot). `is not None` so an intentionally empty
+            # published value is preserved rather than replaced by the graph's.
+            name=agent.name if agent.name is not None else graph.name,
+            description=(
+                agent.description
+                if agent.description is not None
+                else graph.description
+            ),
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/model.py:361]: Fixed in cd4d995e0: `list_library_agents` search now also matches the snapshotted `LibraryAgent.name`/`description`, so the displayed marketplace title is searchable. Added `test_list_library_agents_search_matches_snapshot_and_graph`.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py [hunk 13]
```diff
@@ -26,11 +26,14 @@ async def resolve_graph_for_library(
...
     admin: bool,
-) -> GraphModel:
+) -> tuple[GraphModel, prisma.models.StoreListingVersion]:
     """Look up a StoreListingVersion and resolve its graph.
 
     When ``admin=True``, uses ``get_graph_as_admin`` to bypass the marketplace
     APPROVED-only check.  Otherwise uses the regular ``get_graph``.
+
+    Returns the resolved graph together with the StoreListingVersion, so callers
+    can snapshot marketplace metadata without re-querying it.
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py:36]: Not an exposure introduced by this PR. The non-admin path resolves graph access via `get_graph(user_id=...)`, which grants a non-owner access only through a `StoreListingVersion` with `submissionStatus == APPROVED` and `isDeleted == False`, gated on the same graph version (backend/data/graph.py). A pending/rejected/deleted listing therefore raises `NotFoundError` before any snapshot happens. This PR does not change that lookup or authorization — it only additionally reads cosmetic name/description/image fields from the SLV that `resolve_graph_for_library` already fetches. Resolving as a false positive for this change.

File: autogpt_platform/backend/backend/api/features/library/_add_to_library.py [hunk 14]
```diff
@@ -52,13 +55,33 @@ async def resolve_graph_for_library(
...
+    downloaded agent shows up in the library exactly as it appears in the
+    marketplace.
+    """
+    return {
+        "name": store_listing_version.name,
+        "description": store_listing_version.description,
+        "imageUrl": (
+            store_listing_version.imageUrls[0]
+            if store_listing_version.imageUrls
+            else None
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/_add_to_library.py:76]: `imageUrls` are creator-supplied and already stored on `StoreListingVersion` and rendered as `` throughout the marketplace UI today; snapshotting the same URL onto the library card adds no new sink. A trusted-origin/HTTPS allowlist belongs at submission time and would be a separate, broader change — out of scope for this fix.

File: autogpt_platform/backend/backend/api/features/library/db_test.py [hunk 15]
```diff
@@ -133,6 +133,39 @@ async def test_list_library_agents_is_hidden_filter(
...
+        "OR"
+    ]
+    assert {
+        "description": {"contains": "Published Title", "mode": "insensitive"}
+    } in where["OR"]
+    assert {
+        "AgentGraph": {
+            "is": {"name": {"contains": "Published Title", "mode": "insensitive"}}
+        }
+    } in where["OR"]
```
[the remark this answers is not in the document]
Comment (Abhi1992002) [reply autogpt_platform/backend/backend/api/features/library/db_test.py:166]: Fixed in 6a970731a: added the fourth assertion for the `AgentGraph.is.description` branch so all four search OR-clauses are covered.

**Comment by Abhi1992002** [comment]
All review threads have been addressed and resolved. Summary of changes since the first push:

- **Removed the redundant `StoreListingVersion` fetch** (91a828f0f): `resolve_graph_for_library` now returns the listing it already fetched, and `add_graph_to_library` snapshots from it — no second `find_unique`. This also restored `db_test.py::test_add_agent_to_library`'s single-fetch assertion (the CI failure).
- **Preserve empty published values** (91a828f0f): `from_db` uses `x if x is not None else graph.x`, so an intentionally-empty marketplace name/description is kept rather than falling back to the graph.
- **Search discoverability** (cd4d995e0): `list_library_agents` search now also matches the snapshotted `LibraryAgent.name`/`description`, so a downloaded agent is findable by the title shown on its card.
- **Tests** (91a828f0f, 6a970731a): added coverage for the empty-`imageUrls` branch, empty-string preservation, snapshot-vs-graph fallback, and all four search OR-clauses.

On the two security findings: the non-admin add-to-library path already resolves graph access through `get_graph(user_id=...)`, which grants a non-owner access only via a `StoreListingVersion` that is `APPROVED` and non-deleted, gated on the same graph version — a pending/rejected listing raises `NotFoundError` before any snapshot. This PR doesn't change that authorization; it only reads cosmetic name/description/image fields from the already-resolved listing. Image-URL origin validation is a pre-existing, submission-time concern (these URLs are already stored and rendered in the marketplace UI) and out of scope here.

All checks are green (backend `test (3.11/3.12/3.13)`, e2e, lint, types, migrations).

**Comment by Abhi1992002** [comment]
!deploy

**Comment by Abhi1992002** [comment]
!deploy

**Comment by Pwuts** [comment]
✅ **Preview environment is live** (all services healthy)

- **Deployed:** `2a0f3006ab9d1abd4bd6daf761883a354f70eb83` at 2026-07-28 12:37 UTC
- **Database:** isolated Supabase branch `pr-13680` (state persists across redeploys unless migration drift forces a reset)
- **URLs:** posted in the team Discord

Push more commits, then comment `!deploy` to update · `!undeploy` or close the PR to tear down.

**Review by 0ubbe — APPROVED** [review]
